File: python/src/nnabla/backward_function/concatenate.py
# Copyright 2019,2020,2021 Sony Corporation.
# Copyright 2021 Sony Group Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import nnabla as nn
import nnabla.function as _F
import nnabla.functions as F
from nnabla.function import PythonFunction


class ConcatenateDataGrad(PythonFunction):

    def __init__(self, ctx, axis=None):
        super(ConcatenateDataGrad, self).__init__(ctx)
        self._func = _F.Concatenate(ctx, axis=axis)

    @property
    def name(self):
        return self.__class__.__name__

    @property
    def args(self):
        return self._func.args

    def _create_fwd_inputs_outputs(self, inputs, outputs):
        # fwd_inputs: x_1, ..., x_n
        # fwd_outputs: y
        oshape = inputs[0].shape
        inputs_fwd = []
        for xshape in self.xshapes:
            input_fwd = nn.Variable(xshape, need_grad=True)
            inputs_fwd.append(input_fwd)
        outputs_fwd = [nn.Variable(oshape)]
        return inputs_fwd, outputs_fwd

    def min_inputs(self):
        return 1

    def min_outputs(self):
        return len(self.xshapes)

    def grad_depends_output_data(self, i, o):
        return False

    def grad_depends_input_data(self, i, j):
        return False

    @property
    def xshapes(self):
        return self._xshapes

    @xshapes.setter
    def xshapes(self, xshapes):
        self._xshapes = xshapes

    def setup_impl(self, inputs, outputs):
        # inputs:  dy
        # outputs: dx_1, ..., dx_n
        inputs_fwd, outputs_fwd = self._create_fwd_inputs_outputs(
            inputs, outputs)
        self._func.setup(inputs_fwd, outputs_fwd)
        for i, xshape in enumerate(self.xshapes):
            outputs[i].reset_shape(xshape, True)

    def forward_impl(self, inputs, outputs):
        inputs_fwd, outputs_fwd = self._create_fwd_inputs_outputs(
            inputs, outputs)
        dy = inputs[0].data
        dx_list = [outputs[i].data for i in range(len(self.xshapes))]

        vy = outputs_fwd[0]
        vx_list = inputs_fwd

        vy.grad = dy
        for dx, vx in zip(dx_list, vx_list):
            vx.grad = dx
        self._func.backward(inputs_fwd, outputs_fwd, [
                            False] * len(self.xshapes))

    def backward_impl(self, inputs, outputs, propagate_down, accum):
        if not propagate_down[0]:
            return

        gdy = inputs[0].grad
        gdx_list = [output.grad for output in outputs]

        inputs_fwd, outputs_fwd = self._create_fwd_inputs_outputs(
            inputs, outputs)
        vx_list = inputs_fwd
        vy = outputs_fwd[0]

        for gdx, vx in zip(gdx_list, vx_list):
            vx.data = gdx
        if accum[0]:
            self._func.forward(inputs_fwd, outputs_fwd)
            gdy += vy.data
        else:
            vy.data = gdy
            self._func.forward(inputs_fwd, outputs_fwd)


def concatenate_backward(grad_inputs, inputs, input_shapes, outputs, output_shapes, axis=None):
    """
    Args:
      grad_inputs (list of :obj:`nnabla.Variable`): Propagated grads to this backward function.
      inputs (list of :obj:`nnabla.Variable` and None): Input Variables of the forward function
          if this backward function depends on it. Otherwise, None is set instead.
      input_shapes (list of tuple of :obj:`int`): Input shapes of the forward function.
          The shapes of the inputs in which None is set can be passed.
      outputs (list of :obj:`nnabla.Variable` and None): Output Variables of the forward function
          if this backward function depends on it. Otherwise, None is set instead.
      output_shapes (list of tuple of :obj:`int`): Output shapes of the forward function.
          The shapes of the outputs in which None is set can be passed.
      kwargs (dict of arguments): Dictionary of the corresponding function arguments.

    Return:
      list of Variable: Return the gradients wrt inputs of the corresponding function.
    """
    dy = grad_inputs[0]
    axis = axis if axis is not None else len(dy.shape) - 1
    ctx = nn.get_current_context()
    df = ConcatenateDataGrad(ctx, axis=axis)
    df.xshapes = input_shapes
    dx0 = df(dy)
    return dx0


def concatenate_data_grad_backward(grad_inputs, inputs, input_shapes, outputs, output_shapes, axis=None):
    """
    Args:
      grad_inputs (list of :obj:`nnabla.Variable`): Propagated grads to this backward function.
      inputs (list of :obj:`nnabla.Variable` and None): Input Variables of the forward function
          if this backward function depends on it. Otherwise, None is set instead.
      input_shapes (list of tuple of :obj:`int`): Input shapes of the forward function.
          The shapes of the inputs in which None is set can be passed.
      outputs (list of :obj:`nnabla.Variable` and None): Output Variables of the forward function
          if this backward function depends on it. Otherwise, None is set instead.
      output_shapes (list of tuple of :obj:`int`): Output shapes of the forward function.
          The shapes of the outputs in which None is set can be passed.
      kwargs (dict of arguments): Dictionary of the corresponding function arguments.

    Return:
      list of Variable: Return the gradients wrt inputs of the corresponding function.
    """
    gdx = grad_inputs
    dy = inputs[0]
    axis = axis if axis is not None else len(dy.shape) - 1
    gdy = F.concatenate(*gdx, axis=axis)
    return gdy

File: src/nbla/function/generic/concatenate.cpp
// Copyright (c) 2017 Sony Corporation. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// concatenate.cpp

#include <nbla/array.hpp>
#include <nbla/function/concatenate.hpp>
#include <nbla/variable.hpp>

namespace nbla {

NBLA_REGISTER_FUNCTION_SOURCE(Concatenate, int);

template <typename T>
void Concatenate<T>::setup_impl(const Variables &inputs,
                                const Variables &outputs) {
  Shape_t shape = inputs[0]->shape();
  if (axis_ < 0)
    axis_ += shape.size();

  NBLA_CHECK(axis_ >= 0, error_code::value,
             "axis must not be less than zero, got %d", axis_);
  auto axis = static_cast<Shape_t::size_type>(this->axis_);
  NBLA_CHECK(axis <= shape.size(), error_code::value,
             "axis must be less than or equal to ndim of input. "
             "axis: %d > ndim of input: %d.",
             axis_, shape.size());
  inner_total_size_ = 0;
  for (Variables::size_type i = 0; i < inputs.size(); i++) {
    NBLA_CHECK(inputs[i]->shape().size() != 0, error_code::value,
               "input value(inputs[%d]) does not exist. "
               "inputs[%d]->shape().size(): %d.",
               i, i, inputs[i]->shape().size());
    const int inner_size = inputs[i]->size(this->axis_);
    inner_total_size_ += inner_size;
    if (i >= 1) {
      shape[axis_] += inputs[i]->shape()[axis_]; // Accumulate size along axis
      for (Shape_t::size_type j = 0; j < shape.size(); j++) {
        if (j != axis) {
          NBLA_CHECK(inputs[i]->shape()[j] == shape[j], error_code::value,
                     "Dimensions of inputs must match. "
                     "inputs[%d]->shape()[%d]: %d != "
                     "inputs[0]->shape()[%d]: %d.",
                     i, j, inputs[i]->shape()[j], j, shape[j]);
        }
      }
    }
  }
  outputs[0]->reshape(shape, true);
  outer_size_ = inputs[0]->size() / inputs[0]->size(axis_);
}

template <class T>
void Concatenate<T>::forward_impl(const Variables &inputs,
                                  const Variables &outputs) {
  T *y = outputs[0]->cast_data_and_get_pointer<T>(this->ctx_, true);
  int inner_offset = 0;
  for (Variables::size_type c = 0; c < inputs.size(); ++c) {
    const T *x = inputs[c]->get_data_pointer<T>(this->ctx_);
    const int inner_size = inputs[c]->size(this->axis_);
    for (int o = 0; o < outer_size_; ++o) {
      for (int i = 0; i < inner_size; ++i) {
        y[o * inner_total_size_ + inner_offset + i] = x[o * inner_size + i];
      }
    }
    inner_offset += inner_size;
  }
}

template <class T>
void Concatenate<T>::backward_impl(const Variables &inputs,
                                   const Variables &outputs,
                                   const vector<bool> &propagate_down,
                                   const vector<bool> &accum) {
  const T *dy = outputs[0]->get_grad_pointer<T>(this->ctx_);
  int inner_offset = 0;

  for (Variables::size_type c = 0; c < inputs.size(); ++c) {
    const int inner_size = inputs[c]->size(this->axis_);
    if (propagate_down[c]) {
      T *dx = inputs[c]->cast_grad_and_get_pointer<T>(this->ctx_, !accum[c]);
      for (int o = 0; o < outer_size_; ++o) {
        for (int i = 0; i < inner_size; ++i) {
          T &rdx = dx[o * inner_size + i];
          rdx = (accum[c] ? rdx : (T)0) +
                dy[o * inner_total_size_ + inner_offset + i];
        }
      }
    }
    inner_offset += inner_size;
  }
}
}
